### ISSUE TITLE
Could you pls update the pysam to make it compatible with python version > 3.8.0

### ISSUE THREAD

**Issue by yangyxt** [pysam-developers/pysam#924]
I use pysam with the multiprocessing. Where I always got an error indicating the struct.error() (https://stackoverflow.com/questions/47776486/python-struct-error-i-format-requires-2147483648-number-2147483647).

The new version python 3.8.0 has fixed up this issue where I can't install pysam in python 3.8.0. Pls consider updating pysam so it can be compatible with the highest python version. Thanks

**Comment by MrOlm** [comment]
As a +1, I would also appreciate the ability to use pysam with version 3.8 for it's ability to use shared memory when multiprocessing

**Comment by jamestwebber** [comment]
Python 3.7 has 9 months before it will no longer receive security updates [see here](https://endoflife.date/python). It would be great if `pysam` supported 3.8 (or higher) before then :)

**Comment by jmarshall** [comment]
PyPI contains [pysam wheels](https://pypi.org/project/pysam/#files) for Python 3.6 through 3.10. I build pysam regularly locally with Python 3.10 without trouble, though admittedly I do not use it in conjunction with `multiprocessing`.

Is the problem you encounter limited to use in conjunction with `multiprocessing`? Please provide a sample program that has problems with recent Python versions due to pysam.

**Comment by jamestwebber** [comment]
Yeah it's specifically an issue with `multiprocessing`: python 3.8+ is unable to pickle things, which it needs to pass them between processes. I'll try to make a minimal example in a bit.

**Comment by jmarshall** [comment]
It will be interesting to see which classes you are trying to pickle.

#752 claimed there was an issue with `AlignedSegment`, which really ought to work. OTOH #767 wanted to pickle `AlignmentFile`, which is very unlikely to be feasible.

**Comment by jamestwebber** [comment]
In our case it looks like the issue was in trying to pass an `AlignmentHeader` to a subprocess–this raises the error `TypeError: no default __reduce__ due to non-trivial __cinit__` (searching for that is how I found this issue).

I think we can work around this by passing a dictionary instead, but it's odd that older versions are fine with passing the header object but 3.8+ is not.

Somewhat minimal example follows. This just echos the bam file. Works on python 3.7.12, not on python 3.8. Passing the header as a dictionary fixes this, but I imagine there are more complex situations where the fix is not so easy.

<details>

```python

import click

import pysam
import multiprocessing as mp


@click.command(name="example")
@click.argument("input-bam", type=click.File("rb"))
@click.argument("output-bam", type=click.Path(exists=False))
def main(input_bam, output_bam):
    manager = mp.Manager()
    results = manager.Queue()

    with pysam.AlignmentFile(input_bam, "rb", check_sq=False) as bam_file:
        # Start output worker:
        output_worker = mp.Process(
            target=_write_thread_fn,
            args=(results, bam_file.header, output_bam)
        )
        output_worker.start()

        for r in bam_file:
            results.put(r.to_string())

        results.put(None)
        output_worker.join()


def _write_thread_fn(out_queue, out_bam_header, out_bam_file_name):
    # out_bam_header = pysam.AlignmentHeader.from_dict(out_bam_header)
    with pysam.AlignmentFile(
        out_bam_file_name, "wb", header=out_bam_header
    ) as out_bam_file:
        while True:
            read = out_queue.get()
            if read is None:
                break

            read = pysam.AlignedSegment.fromstring(read, out_bam_header)
            out_bam_file.write(read)
```
</details>
